### ISSUE TITLE
Compliation time grows exponentially

### ISSUE THREAD

**Issue by pclayton** [polyml/polyml#138]
I am recording an issue reported to David a long time ago in the email [Re: 'make compiler' resulting in slow compilation.eml.gz](https://github.com/polyml/polyml/files/5000188/Re.make.compiler.resulting.in.slow.compilation.eml.gz).

This [example](https://github.com/polyml/polyml/files/5000386/polyml_compiler_issue-201309829.sml.gz) appears to show that the time taken to compile a function grows exponentially with respect to the size of the function.  The test cases produce the following timing results with Poly/ML 5.8.1:
```
Timing - compile: 0.4 run: 0.0
Timing - compile: 0.7 run: 0.0
Timing - compile: 1.6 run: 0.0
Timing - compile: 3.2 run: 0.0
Timing - compile: 6.7 run: 0.0
Timing - compile: 14.7 run: 0.0
```
As noted in the original message, the space usage may also be growing exponentially - perhaps there is a path explosion in the generated code.

In summary, given the code
```
fun C f =
  fn
    SOME n => f n
  | NONE   => f 0

infixr 1 &&&>
fun (f &&&> g) h (a, b) = f (fn a' => g (fn b' => h (a', b')) b) a
```
the examples have the form
```
fun test<N> f =
  (
    C
     &&&> C
     &&&> C
...
     &&&> C
     &&&> C
  )
    f
;
```

**Comment by dcjm** [comment]
Thanks for reminding me about this.  There was a blow-up in the inline expansion which turned out to be simpler to fix than I feared.  Seems to have been fixed with 8e4d341.

**Comment by pclayton** [comment]
Thanks - this has fixed it.  I see a 4-5% improvement in load times:
5.8	user	0m34.696s
5.8.1	user	0m34.259s
d471c5e	user	0m32.895s
f54aa41	user	0m32.797s
(each time is best of 3 runs)

**Comment by pclayton** [comment]
Thanks - this has fixed it.  I see a 4-5% improvement in load times:
5.8	user	0m34.696s
5.8.1	user	0m34.259s
d471c5e	user	0m32.895s
f54aa41	user	0m32.797s
(each time is best of 3 runs)

**Comment by pclayton** [comment]
Thanks - this has fixed it.  I see a 4-5% improvement in load times:
5.8	user	0m34.696s
5.8.1	user	0m34.259s
d471c5e	user	0m32.895s
f54aa41	user	0m32.797s
(each time is best of 3 runs)
